### ISSUE TITLE
Registering the same provider with papiea-engine twice sometimes fails w/HTTP 500

### ISSUE THREAD

**Issue by joshua-berry-ntnx** [nutanix/papiea#381]
Sometimes (but not always), registering the BMaaS provider with papiea-engine fails. From bmaas/papiea-sdk:

```
bmaas     | 5/7/2020, 6:47:30 PM:  Error: Request failed with status code 500
bmaas     |     at createError (/app/node_modules/axios/lib/core/createError.js:16:15)
bmaas     |     at settle (/app/node_modules/axios/lib/core/settle.js:18:12)
... {
...
bmaas     |   config: {
...
bmaas     |     method: 'post',
bmaas     |     baseURL: 'http://bmaas_papiea:3000/provider',
bmaas     |     url: 'http://bmaas_papiea:3000/provider/',
bmaas     |     data: `{"kinds":[{"name":"machine-type","name_plural":"machine-types",...}`
bmaas     |   },
bmaas     |   request: <ref *1> ClientRequest {
...
bmaas     |     _header: 'POST /provider/ HTTP/1.1\r\n' +
bmaas     |       'Accept: application/json, text/plain, */*\r\n' +
bmaas     |       'Content-Type: application/json\r\n' +
bmaas     |       'Authorization: Bearer IAmTheAdmin\r\n' +
bmaas     |       'User-Agent: axios/0.18.1\r\n' +
bmaas     |       'Content-Length: 10542\r\n' +
bmaas     |       'Host: bmaas_papiea:3000\r\n' +
bmaas     |       'Connection: close\r\n' +
bmaas     |       '\r\n',
...
bmaas     |     method: 'POST',
...
bmaas     |     path: '/provider/',
...
bmaas     |   },
bmaas     |   response: {
bmaas     |     status: 500,
bmaas     |     statusText: 'Internal Server Error',
bmaas     |     headers: {
bmaas     |       'x-powered-by': 'Express',
bmaas     |       'content-type': 'application/json; charset=utf-8',
bmaas     |       'content-length': '481',
bmaas     |       etag: 'W/"1e1-gjNpBnQGUykVmfY2c7hNGKv4AoM"',
bmaas     |       date: 'Thu, 07 May 2020 18:47:30 GMT',
bmaas     |       connection: 'close'
bmaas     |     },
...
bmaas     |     data: { error: [Object] }
bmaas     |   }
bmaas     | }
```

And here is what I see from the engine and mongo:

```
bmaas_mongo | 2020-05-07T18:47:30.681+0000 I  SHARDING [conn8] Marking collection papiea.intentful_kinds as collection version: <unsharded>
bmaas_papiea | Default handle got error: E11000 duplicate key error collection: papiea.intentful_kinds index: provider_prefix_1_provider_version_1_kind_name_1 dup key: { provider_prefix: "bmaas", provider_version: "0.0.1", kind_name: "machine" }
bmaas_papiea | {"0":"BulkWriteError: E11000 duplicate key error collection: papiea.intentful_kinds index: provider_prefix_1_provider_version_1_kind_name_1 dup key: { provider_prefix: \"bmaas\", provider_version: \"0.0.1\", kind_name: \"machine\" }\n    at OrderedBulkOperation.handleWriteError (/code/papiea-engine/node_modules/mongodb/lib/bulk/common.js:1112:11)\n    at resultHandler (/code/papiea-engine/node_modules/mongodb/lib/bulk/common.js:501:23)\n    at /code/papiea-engine/node_modules/mongodb-core/lib/connection/pool.js:397:18\n    at process._tickCallback (internal/process/next_tick.js:61:11)","level":"error","message":{"error":{"code":500,"errors":[{"message":"E11000 duplicate key error collection: papiea.intentful_kinds index: provider_prefix_1_provider_version_1_kind_name_1 dup key: { provider_prefix: \"bmaas\", provider_version: \"0.0.1\", kind_name: \"machine\" }"}],"message":"E11000 duplicate key error collection: papiea.intentful_kinds index: provider_prefix_1_provider_version_1_kind_name_1 dup key: { provider_prefix: \"bmaas\", provider_version: \"0.0.1\", kind_name: \"machine\" }"}},"timestamp":"2020-05-07T18:47:30.734Z"}
```

Simply restarting the engine/mongo does not fix it; I have to blow away the temporary volumes created by `docker-compose` for mongo and go back to a pristine state before I hit this again.

Normally I would anticipate this sort of thing to happen because of some schema change in the provider. However, there are no such changes here—the BMaaS code is exactly the same between the first (successful) registration and subsequent unsuccessful registrations.

**Comment by joshua-berry-ntnx** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Papiea bugs - needed for inventory](https://trello.com/c/R4fVY0ra/94-papiea-bugs-needed-for-inventory)

**Comment by joshua-berry-ntnx** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Address Papiea issues needed for inventory-only BMaaS](https://trello.com/c/V6gmTU73/66-address-papiea-issues-needed-for-inventory-only-bmaas)

**Comment by Emulebest** [comment]
This is resolved by PR #410
